### DISCUSSION TITLE
Supporting multiple worktrees

### DISCUSSION THREAD

**Discussion by wmmc88** [microsoft/git#682] [Q&A]
My existing git workflow is to have multiple git worktrees for the same repo, within a bare clone of the repo. 
ex. 
`repo/worktree-folder/branch1`
`repo/worktree-folder/branch2`
`repo/worktree-folder/branch3`
`repo/.git`

and I create/delete worktrees as needed. `scalar` seems to have a single git worktree that exists at `src` for a given enlistment. Is there a recommended way to get `scalar` to support multiple worktrees per enlistment?

**Comment by dscho** [comment]
How about running [`scalar register`](https://git-scm.com/docs/scalar#_register) in one of the worktrees? That _should_ accomplish what you seek.

**Reply by derrickstolee** [reply]
I think the concern is about how `git worktree` fits with `scalar clone` (and specifically with `microsoft/git`). Theoretically, they work just fine. However, be careful about the `src` directory. `scalar clone` creates `<repo>/src` to allow build artifacts from the repo to be placed outside of the working directory. This helps speed up Git commands as well as other build operations. But if you use `git worktree add ../src2`, then those build outputs will start thrashing on each other.

**Reply by wmmc88** [reply]
I think as it stands with `scalar clone`'s current file structure, it seems that one enlistment can only have one worktree to avoid thrashing from the out-of-tree build artifacts. I think if `scalar clone` were able to add one additional layer of folder structures per worktree, it would be easier to support this multiple worktrees in a single enlistment scenario.

**Reply by dscho** [reply]
Assuming that your project indeed puts the build products into a directory tree that is adjacent to the Git worktree (e.g. into `repo/worktree-folder/out` if the worktree is `repo/worktree-folder/branch1`), an easy way to fix this would be to initialize the worktrees into slightly different paths (note the extra `/src`):

- repo/worktree-folder/branch1/src
- repo/worktree-folder/branch2/src
- repo/worktree-folder/branch3/src

That way, the respective `out` folders would be in different folders and would therefore not clash.

If the worktrees have already been initialized, you can use [`git worktree move`](https://git-scm.com/docs/git-worktree#Documentation/git-worktree.txt-move) to change the location; You will need to do this individually for every worktree.
